OpenRA.Game/Activities/Activity.cs

```diff
@@ -19,122 +19,26 @@ namespace OpenRA.Activities
-	 * - Do not "reuse" (with "SequenceActivities", for example) activity objects that have already finished running.
+	 * - Do not use "return new SomeActivity()" as that will break the queue. Queue the new activity and use "return NextActivity" instead.
+	 * - Do not "reuse" (with "SequenceActivities", for example) activity objects that have already started running.
 	 *   Queue a new instance instead.
 	 * - Avoid calling actor.CancelActivity(). It is almost always a bug. Call activity.Cancel() instead.
-	 * - Do not return the Parent explicitly unless you have an extremly good reason. "return NextActivity"
-	 *   will do the right thing in all circumstances.
-	 * - You do not need to care about the ChildActivity pointer advancing through the list of children,
-	 *   the activity code already takes care of that.
-	 * - If you want to check whether there are any follow-up activities queued, check against "NextInQueue"
-	 *   in favour of "NextActivity" to avoid checking against the Parent activity.
 	 */
 	public abstract class Activity
 	{
 		public ActivityState State { get; private set; }
 
-		/// <summary>
-		/// Returns the top-most activity *from the point of view of the calling activity*. Note that the root activity
-		/// can and likely will have next activities of its own, which would in turn be the root for their children.
-		/// </summary>
-		public Activity RootActivity
-		{
-			get
-			{
-				var p = this;
-				while (p.ParentActivity != null)
-					p = p.ParentActivity;
-
-				return p;
-			}
-		}
-
-		Activity parentActivity;
-		public Activity ParentActivity
-		{
-			get
-			{
-				return parentActivity;
-			}
-
-			protected set
-			{
-				parentActivity = value;
-
-				var next = NextInQueue;
-				if (next != null)
-					next.ParentActivity = parentActivity;
-			}
-		}
-
 		Activity childActivity;
 		protected Activity ChildActivity
 		{
-			get
-			{
-				return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null;
-			}
-
-			set
-			{
-				if (value == this || value == ParentActivity || value == NextInQueue)
-					childActivity = null;
-				else
-				{
-					childActivity = value;
-
-					if (childActivity != null)
-						childActivity.ParentActivity = this;
-				}
-			}
-		}
-
-		Activity nextActivity;
-
-		/// <summary>
-		/// The getter will return either the next activity or, if there is none, the parent one.
-		/// </summary>
-		public virtual Activity NextActivity
-		{
-			get
-			{
-				return nextActivity != null ? nextActivity : ParentActivity;
-			}
-
-			set
-			{
-				if (value == this || value == ParentActivity || (value != null && value.ParentActivity == this))
-					nextActivity = null;
-				else
-				{
-					nextActivity = value;
-
-					if (nextActivity != null)
-						nextActivity.ParentActivity = ParentActivity;
-				}
-			}
+			get { return childActivity != null && childActivity.State < ActivityState.Done ? childActivity : null; }
+			set { childActivity = value; }
```

Comment:
It might be a good idea to check and throw an exception if `Game.Settings.Debug.StrictActivityChecking && childActivity != null && childActivity.State < ActivityState.Done`.

Something for a followup PR, perhaps.

Reply:
In theory, this could be removed in the future as well. The `ChildActivity = ActivityUtils.RunActivity(...)` pattern should see to it that ChildActivity gets set to null eventually, and it shouldn't ever be left in the Done state. I left it in out of caution for now.